pkg/operators/uidgidresolver/usergroupcache.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uidgidresolver
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/cachedmap"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/utils/host"
+)
+
+// UserGroupCache is a cache of user names, uids, group names and gids
+type UserGroupCache interface {
+	Start() error
+	Stop()
+
+	GetUsername(uint32) string
+	GetGroupname(uint32) string
+}
+
+type userGroupCache struct {
+	userCache  cachedmap.CachedMap[uint32, string]
+	groupCache cachedmap.CachedMap[uint32, string]
+
+	loopFinished  chan struct{}
+	useCount      int
+	useCountMutex sync.Mutex
+	watcher       *fsnotify.Watcher
+}
+
+const (
+	passwdFileName = "passwd"
+	groupFileName  = "group"
+	baseDirPath    = "/etc"
+)
+
+var (
+	userGroupCacheSingleton *userGroupCache
+	userGroupOnce           sync.Once
```

Comment:
nit: you can use [OnceValue](https://cs.opensource.google/go/go/+/go1.22.2:src/sync/oncefunc.go;l=43) in go 1.21 to replace these two.